File: Go/Go_1000-/1656.go
// Go
// leetcode 1656
// https://leetcode.cn/problems/design-an-ordered-stream/

type OrderedStream struct {
	arr []int
	ptr int
	tmp []byte
}


func Constructor(n int) OrderedStream {
	arr := make([]int, n)
	for i := 0; i < n; i++ {
		arr[i] = -1
	}
	return OrderedStream{arr, 0, make([]byte, 5)}
}

func getIdx(s string) int {
	idx := 0
	for i := 0; i < 5; i++ {
		idx = idx * 26 + int(s[i] - 'a')
	}
	return idx
}

func (t *OrderedStream) LoadInTmp(x int) bool {
	if x == -1 {
		return false
	}

	for i := 4; i >= 0; i-- {
		t.tmp[i] = 'a' + byte(x % 26)
		x /= 26
	}

	return true
}

func (t *OrderedStream) Insert(idKey int, value string) []string {
	res := []string{}
	t.arr[idKey - 1] = getIdx(value)
	p := t.ptr
	for p < len(t.arr) {
		if t.LoadInTmp(t.arr[p]) {
			res = append(res, string(t.tmp))
			p++
		} else {
			break
		}
	}
	t.ptr = p
	return res
}


/**
 * Your OrderedStream object will be instantiated and called as such:
 * obj := Constructor(n);
 * param_1 := obj.Insert(idKey,value);
 */

File: cpp/cpp_1000-/1656.cc
// C++
// leetcode 1656
// https://leetcode.cn/problems/design-an-ordered-stream/

class OrderedStream {
public:
    vector<int> vs;
    int ptr;
    char tmp[5];
    OrderedStream(int n) {
        vs.resize(n, -1);
        ptr = 0;
    }
    
    int getIdx(string &s) {
        int idx = 0, u = -1;
        for (int i = 0; i < 5; i++) {
            u = s[i] - 'a';
            idx = idx * 26 + u;
        }
        return idx;
    }

    bool loadInTmp(int x) {
        if (x == -1)
            return false;

        char c;
        for (int i = 4; i >= 0; i--) {
            c = 'a' + (x % 26);
            tmp[i] = c;
            x /= 26;
        }
        return true;
    }

    vector<string> insert(int idKey, string value) {
        vector<string> res;
        vs[idKey - 1] = getIdx(value);
        int p = ptr;
        while (p < vs.size()) {
            if (loadInTmp(vs[p])) {
                res.push_back(string(tmp, tmp + 5));
                p++;
            } else {
                break;
            }
        }
        ptr = p;
        return res;
    }
};

/**
 * Your OrderedStream object will be instantiated and called as such:
 * OrderedStream* obj = new OrderedStream(n);
 * vector<string> param_1 = obj->insert(idKey,value);
 */